File: public/files/js/stores/query/filter.ts
/*
 * Copyright (c) 2016 Charles University in Prague, Faculty of Arts,
 *                    Institute of the Czech National Corpus
 * Copyright (c) 2016 Tomas Machalek <tomas.machalek@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2
 * dated June, 1991.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

/// <reference path="../../types/common.d.ts" />
/// <reference path="../../types/ajaxResponses.d.ts" />
/// <reference path="../../../ts/declarations/immutable.d.ts" />
/// <reference path="../../../ts/declarations/cqlParser.d.ts" />


import * as Immutable from 'vendor/immutable';
import {SimplePageStore} from '../../util';
import {PageModel} from '../../tpl/document';
import {MultiDict} from '../../util';
import {parse as parseQuery} from 'cqlParser/parser';
import {TextTypesStore} from '../textTypes/attrValues';
import {QueryContextStore} from './context';
import {GeneralQueryFormProperties, GeneralQueryStore} from './main';


/**
 * This interface encodes values of multiple filter values. Array indices
 * should match query pipeline with non-filter ones represented by
 * 'undefined'.
 */
export interface FilterFormProperties extends GeneralQueryFormProperties {
    filters:Array<string>;
    maincorps:Array<[string, string]>;
    currQueryTypes:Array<[string, string]>;
    currQueries:Array<[string, string]>;  // current queries values (e.g. when restoring a form state)
    currDefaultAttrValues:Array<[string, string]>;
    tagBuilderSupport:Array<[string, boolean]>;
    currLposValues:Array<[string, string]>;
    currQmcaseValues:Array<[string, boolean]>;
    currInclkwicValues:Array<[string, boolean]>;
    inputLanguage:string;
    currPnFilterValues:Array<[string, string]>;
    currFilflVlaues:Array<[string, string]>;
    currFilfposValues:Array<[string, string]>;
    currFiltposValues:Array<[string, string]>;
    isWithin:boolean;
    opLocks:Array<[string, boolean]>;
}

/**
 *
 */
export type FilterWidgetsMap = Immutable.Map<string, Immutable.List<string>>;

/**
 *
 */
export function fetchFilterFormArgs<T>(args:{[ident:string]:AjaxResponse.ConcFormArgs},
        key:(item:AjaxResponse.FilterFormArgs)=>T):Array<[string, T]> {
    const ans = [];
    for (let formId in args) {
        if (args.hasOwnProperty(formId) && args[formId].form_type === 'filter') {
            ans.push([formId, key(<AjaxResponse.FilterFormArgs>args[formId])]);
        }
    }
    return ans;
}

/**
 * FilterStore handles all the filpublic/files/js/stores/query/replay.tsters applied within a query "pipeline".
 * Each filter is identified by its database ID (i.e. a key used by conc_persistence
 * plug-in to store it). Please note that it does not know the order of filters
 * in pipeline (it is up to QueryReplay store to handle this).
 */
export class FilterStore extends GeneralQueryStore implements Kontext.QuerySetupHandler {

    private maincorps:Immutable.Map<string, string>;

    private queries:Immutable.Map<string, string>;

    private queryTypes:Immutable.Map<string, string>;

    private lposValues:Immutable.Map<string, string>;

    private matchCaseValues:Immutable.Map<string, boolean>;

    private defaultAttrValues:Immutable.Map<string, string>;

    private pnFilterValues:Immutable.Map<string, string>;

    private filflValues:Immutable.Map<string, string>;

    private filfposValues:Immutable.Map<string, string>;

    private filtposValues:Immutable.Map<string, string>;

    private inclkwicValues:Immutable.Map<string, boolean>;

    private tagBuilderSupport:Immutable.Map<string, boolean>;

    /**
     * If true for a certain key then the operation cannot be edited.
     * (this applies e.g. for filters generated by manual line
     * selection).
     */
    private opLocks:Immutable.Map<string, boolean>;

    private inputLanguage:string;

    private isWithin:boolean;

    constructor(dispatcher:Dispatcher.Dispatcher<any>, pageModel:PageModel, textTypesStore:TextTypesStore,
            queryContextStore:QueryContextStore, props:FilterFormProperties) {
        super(dispatcher, pageModel, textTypesStore, queryContextStore, props);

        this.maincorps = Immutable.Map<string, string>(props.maincorps);
        this.queries = Immutable.Map<string, string>(props.currQueries);
        this.queryTypes = Immutable.Map<string, string>(props.currQueryTypes);
        this.lposValues = Immutable.Map<string, string>(props.currLposValues);
        this.matchCaseValues = Immutable.Map<string, boolean>(props.currQmcaseValues);
        this.defaultAttrValues = Immutable.Map<string, string>(props.currDefaultAttrValues);
        this.pnFilterValues = Immutable.Map<string, string>(props.currPnFilterValues);
        this.filflValues = Immutable.Map<string, string>(props.currFilflVlaues);
        this.filfposValues = Immutable.Map<string, string>(props.currFilfposValues);
        this.filtposValues = Immutable.Map<string, string>(props.currFiltposValues);
        this.inclkwicValues = Immutable.Map<string, boolean>(props.currInclkwicValues);
        this.tagBuilderSupport = Immutable.Map<string, boolean>(props.tagBuilderSupport);
        this.opLocks = Immutable.Map<string, boolean>(props.opLocks);


        this.inputLanguage = props.inputLanguage;
        this.currentAction = 'filter_form';
        this.isWithin = props.isWithin;

        this.dispatcher.register((payload:Kontext.DispatcherPayload) => {
            switch (payload.actionType) {
                case 'FILTER_QUERY_INPUT_SELECT_TYPE':
                    this.queryTypes = this.queryTypes.set(payload.props['sourceId'], payload.props['queryType']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_INPUT_SET_QUERY':
                    this.queries = this.queries.set(payload.props['sourceId'], payload.props['query']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_INPUT_APPEND_QUERY':
                    const currQuery = this.queries.get(payload.props['sourceId'])
                    const newQuery =  currQuery + (currQuery && payload.props['prependSpace'] ? ' ' : '') + payload.props['query'];
                    this.queries = this.queries.set(payload.props['sourceId'], newQuery);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_INPUT_SET_LPOS':
                    this.lposValues = this.lposValues.set(payload.props['sourceId'], payload.props['lpos']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_INPUT_SET_MATCH_CASE':
                    this.matchCaseValues = this.matchCaseValues.set(payload.props['sourceId'], payload.props['value']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_INPUT_SET_DEFAULT_ATTR':
                    this.defaultAttrValues = this.defaultAttrValues.set(payload.props['sourceId'], payload.props['value']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_SET_POS_NEG':
                    this.pnFilterValues = this.pnFilterValues.set(payload.props['filterId'], payload.props['value']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_SET_FILFL':
                    this.filflValues = this.filflValues.set(payload.props['filterId'], payload.props['value']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_SET_RANGE':
                    if (payload.props['rangeId'] === 'filfpos') {
                        this.filfposValues = this.filfposValues.set(payload.props['filterId'], payload.props['value']);

                    } else if (payload.props['rangeId'] === 'filtpos') {
                        this.filtposValues = this.filtposValues.set(payload.props['filterId'], payload.props['value']);
                    }
                    this.notifyChangeListeners();
                break;
                case'FILTER_QUERY_SET_INCL_KWIC':
                    this.inclkwicValues = this.inclkwicValues.set(payload.props['filterId'], payload.props['value']);
                    this.notifyChangeListeners();
                break;
                case 'FILTER_QUERY_APPLY_FILTER':
                    this.submitQuery(payload.props['filterId']);
                    this.notifyChangeListeners();
                break;
            }
        });
    }

    getSubmitUrl(filterId:string):string {
        return this.pageModel.createActionUrl('filter', this.createSubmitArgs(filterId).items());
    }

    /**
     * Synchronize user input values from an external source
     * (typically a server response or a local cache).
     */
    syncFrom(fn:()=>RSVP.Promise<AjaxResponse.FilterFormArgs>):RSVP.Promise<AjaxResponse.FilterFormArgs> {
        return fn().then(
            (data) => {
                const filterId = data.op_key;
                if (data.form_type === 'filter') {
                    this.queries = this.queries.set(filterId, data.query);
                    this.queryTypes = this.queryTypes.set(filterId, data.query_type);
                    this.maincorps = this.queryTypes.set(filterId, data.maincorp);
                    this.pnFilterValues = this.pnFilterValues.set(filterId, data.pnfilter);
                    this.filflValues = this.filflValues.set(filterId, data.filfl);
                    this.filfposValues = this.filfposValues.set(filterId, data.filfpos);
                    this.filtposValues = this.filtposValues.set(filterId, data.filtpos);
                    this.inclkwicValues = this.inclkwicValues.set(filterId, data.inclkwic);
                    this.matchCaseValues = this.matchCaseValues.set(filterId, data.qmcase);
                    this.defaultAttrValues = this.defaultAttrValues.set(filterId, data.default_attr_value);
                    this.tagBuilderSupport = this.tagBuilderSupport.set(filterId, data.tag_builder_support);
                    this.lposValues = this.lposValues.set(filterId, data.lpos);
                    this.opLocks = this.opLocks.set(filterId, false);
                    return data;

                } else if (data.form_type === 'locked') {
                    this.opLocks = this.opLocks.set(filterId, true);
                    return data;

                } else {
                    throw new Error('Cannot sync filter store - invalid form data type: ' + data.form_type);
                }
            }
        );
    }

    private createSubmitArgs(filterId:string):MultiDict {
        const args = this.pageModel.getConcArgs();
        args.replace('pnfilter', [this.pnFilterValues.get(filterId)]);
        args.replace('filfl', [this.filflValues.get(filterId)]);
        args.replace('filfpos', [this.filfposValues.get(filterId)]);
        args.replace('filtpos', [this.filtposValues.get(filterId)]);
        args.replace('inclkwic', [this.inclkwicValues.get(filterId) ? '1' : '0']);
        args.replace('queryselector', [`${this.queryTypes.get(filterId)}row`]);
        if (this.isWithin) {
            args.replace('within', ['1']);

        } else {
            args.remove('within');
        }
        args.replace(this.queryTypes.get(filterId), [this.queries.get(filterId)]);
        return args;
    }

    submitQuery(filterId:string):void {
        const error = this.validateQuery(this.queries.get(filterId), this.queryTypes.get(filterId));
        if (!error || window.confirm(this.pageModel.translate('global__query_type_mismatch'))) {
            const args = this.createSubmitArgs(filterId);
            window.location.href = this.pageModel.createActionUrl('filter', args.items());
        }
    }

    getCorpora():Immutable.List<string> {
        return Immutable.List<string>([this.maincorps]);
    }

    getAvailableAlignedCorpora():Immutable.List<{n:string; label:string}> {
        return Immutable.List<{n:string; label:string}>();
    }

    getQueryTypes():Immutable.Map<string, string> {
        return this.queryTypes;
    }

    getLposValues():Immutable.Map<string, string> {
        return this.lposValues;
    }

    getMatchCaseValues():Immutable.Map<string, boolean> {
        return this.matchCaseValues;
    }

    getDefaultAttrValues():Immutable.Map<string, string> {
        return this.defaultAttrValues;
    }

    getInputLanguage():string {
        return this.inputLanguage;
    }

    getQuery(filterId:string):string {
        return this.queries.get(filterId);
    }

    getPnFilterValues():Immutable.Map<string, string> {
        return this.pnFilterValues;
    }

    getFilfposValues():Immutable.Map<string, string> {
        return this.filfposValues;
    }

    getFiltposValues():Immutable.Map<string, string> {
        return this.filtposValues;
    }

    getFilflValues():Immutable.Map<string, string> {
        return this.filflValues;
    }

    getInclKwicValues():Immutable.Map<string, boolean> {
        return this.inclkwicValues;
    }

    getOpLocks():Immutable.Map<string, boolean> {
        return this.opLocks;
    }

    getSupportedWidgets():FilterWidgetsMap {
        const getWidgets = (filterId:string):Array<string> => {
            switch (this.queryTypes.get(filterId)) {
                case 'iquery':
                case 'lemma':
                case 'phrase':
                case 'word':
                case 'char':
                    return ['keyboard', 'history'];
                case 'cql':
                    const ans = ['keyboard', 'history'];
                    if (this.tagBuilderSupport.get(filterId)) {
                        ans.push('tag');
                    }
                    return ans;
            }
        }
        return Immutable.Map<string, Immutable.List<string>>(
            this.queries.keySeq().map(filterId => {
                return [filterId, Immutable.List<string>(
                getWidgets(filterId))];
            })
        );
    }
}

File: lib/action/argmapping/conc/filter.py
# Copyright(c) 2017 Charles University, Faculty of Arts,
#                   Institute of the Czech National Corpus
# Copyright(c) 2017 Tomas Machalek <tomas.machalek@gmail.com>
# Copyright(c) 2021 Martin Zimandl <martin.zimandl@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 2
# dated June, 1991.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Tuple

import plugins
from action.argmapping import Args
from action.argmapping.conc.base import ConcFormArgs
from action.argmapping.error import ArgumentMappingError, ValidationError
from action.plugin.ctx import AbstractCorpusPluginCtx
from dataclasses_json import dataclass_json
from plugin_types.corparch.corpus import TagsetInfo

from .query import QueryFormArgs
from .base import AbstractRawQueryDecoder


@dataclass_json
@dataclass
class _FilterFormArgs:
    form_type: str = 'filter'
    query_type: str = 'simple'
    query: str = ''
    parsed_query: List[Any] = field(default_factory=list)
    maincorp: str = ''
    pnfilter: str = 'p'
    filfl: str = 'f'
    filfpos: str = '-5'
    filtpos: str = '5'
    inclkwic: bool = True
    qmcase: bool = False
    default_attr: str = 'word'
    use_regexp: bool = False
    has_lemma: bool = False
    tagsets: List[TagsetInfo] = field(default_factory=list)
    within: bool = False
    no_query_history: bool = False


class FilterFormArgs(ConcFormArgs[_FilterFormArgs], AbstractRawQueryDecoder):
    """
    FilterFormArgs provides methods to handle concordance filter
    form arguments represented by the _FilterFormArgs data class.
    """
    @classmethod
    async def create(cls, plugin_ctx: AbstractCorpusPluginCtx, maincorp: str, persist: bool) -> 'FilterFormArgs':
        self = FilterFormArgs(plugin_ctx, maincorp, persist)
        await self._add_corpus_metadata()
        return self

    def __init__(self, plugin_ctx: AbstractCorpusPluginCtx, maincorp: str, persist: bool) -> None:
        super().__init__(persist)
        self._plugin_ctx = plugin_ctx
        self.data = _FilterFormArgs(
            maincorp=maincorp
        )

    def update_by_user_query(self, data):
        self.data.query_type = data['qtype']
        self.data.query = data.get('query')
        self.data.parsed_query = data.get('queryParsed', [])
        self.data.pnfilter = data['pnfilter']
        self.data.filfl = data['filfl']
        self.data.filfpos = data['filfpos']
        self.data.filfpos = data['filfpos']
        self.data.filtpos = data['filtpos']
        self.data.inclkwic = data['inclkwic']
        self.data.qmcase = data['qmcase']
        self.data.within = data.get('within')
        self.data.default_attr = data['default_attr']
        self.data.use_regexp = data.get('use_regexp', False)
        self.data.no_query_history = data.get('no_query_history', False)

    def from_raw_query(self, q, corpname) -> 'FilterFormArgs':
        """
        Parses queries like: p-5 -1 1 [lc="author" & tag="N.*"]
        """
        srch = re.search(r'^([nNpP])(-?\d+)\s+(-?\d+)\s+(-?\d+)\s+(.+)$', q)
        if not srch:
            raise ValueError('unsupported filter expression: {}'.format(q))
        self.data.pnfilter = srch.group(1)
        if srch[0] in ('n', 'p'):
            self.data.inclkwic = True
        else:
            self.data.inclkwic = False
        self.data.filfpos = srch.group(2)
        self.data.filtpos = srch.group(3)
        if srch.group(4) != '1':
            raise ValueError(f'unsupported token selection {srch.group(4)}; only \'1\' is supported')
        self.data.filfl = 'f'
        self.data.query = srch.group(5)
        self.data.query_type = 'advanced'
        return self

    def apply_last_used_opts(self, data: Dict[str, Any]):
        self.data.filfpos = data['filfpos']
        self.data.filtpos = data['filtpos']
        self.data.filfl = data['filfl']
        self.data.inclkwic = data['inclkwic']
        self.data.qmcase = data['qmcase']
        self.data.default_attr = data['default_attr']
        self.data.use_regexp = data['use_regexp']

    async def _add_corpus_metadata(self):
        with plugins.runtime.CORPARCH as ca, plugins.runtime.TAGHELPER as th:
            corp_info = await ca.get_corpus_info(self._plugin_ctx, self.data.maincorp)
            self.data.has_lemma = corp_info.manatee.has_lemma
            self.data.tagsets = [d.to_dict() for d in corp_info.tagsets]
            for tagset in self.data.tagsets:
                tagset['widgetEnabled'] = tagset['widgetEnabled'] and await th.tags_available_for(
                    self._plugin_ctx, self.data.maincorp, tagset['ident'])

    def validate(self):
        try:
            int(self.data.filfpos)
        except ValueError:
            return ValidationError('Invalid value for filfpos: {}'.format(self.data.filfpos))
        try:
            int(self.data.filtpos)
        except ValueError:
            return ValidationError('Invalid value for filtpos: {}'.format(self.data.filtpos))


@dataclass_json
@dataclass
class _SubHitsFilterFormArgs:
    form_type: str = 'subhits'


class SubHitsFilterFormArgs(ConcFormArgs[_SubHitsFilterFormArgs], AbstractRawQueryDecoder):
    """
    SubHitsFilterFormArgs provides methods to handle concordance "sub hits"
    filter form arguments represented by the _SubHitsFilterFormArgs data class.
    """

    def __init__(self, persist: bool) -> None:
        super().__init__(persist)
        self.data = _SubHitsFilterFormArgs()

    def from_raw_query(self, q, corpname) -> 'SubHitsFilterFormArgs':
        if q[1:] != "":
            raise ValueError('Raw SubHitsFilterFormArgs query must be empty')
        return self


@dataclass_json
@dataclass
class _FirstHitsFilterFormArgs:
    doc_struct: str
    form_type: str = 'firsthits'


class FirstHitsFilterFormArgs(ConcFormArgs[_FirstHitsFilterFormArgs], AbstractRawQueryDecoder):
    """
    FirstHitsFilterFormArgs provides methods to handle concordance "first hit in document"
    filter form arguments represented by the _SubHitsFilterFormArgs data class.
    """

    def __init__(self, persist: bool, doc_struct: str) -> None:
        super().__init__(persist)
        self.data = _FirstHitsFilterFormArgs(doc_struct=doc_struct)

    def from_raw_query(self, q, corpname) -> 'FirstHitsFilterFormArgs':
        self.data.doc_struct = q[1:]
        return self

class ContextFilterArgsConv:
    """
    Converts context filter (i.e. the filter which is part of the main query form)
    form arguments into the regular filter ones.
    """

    def __init__(self, plugin_ctx: AbstractCorpusPluginCtx, args: QueryFormArgs) -> None:
        self.plugin_ctx = plugin_ctx
        self.args = args

    @staticmethod
    def _convert_query(attrname: str, items: List[str], fctxtype: str) -> str:
        if fctxtype == 'any':
            return ' | '.join('[{0}="{1}"]'.format(attrname, v) for v in items)
        elif fctxtype == 'all':
            # here we assume len(items) == 1
            # (it's ok - see function append_filter() in _set_first_query action
            # where the operation is split into multiple filters as there
            # is no way how to specify a conjunction in a single query
            return '[{0}="{1}"]'.format(attrname, items[0])
        elif fctxtype == 'none':
            return ' | '.join('[{0}="{1}"]'.format(attrname, v) for v in items)
        raise ValueError(f'Unknown type fctxtype = {fctxtype}')

    async def __call__(self, corpname: str, attrname: str, items: List[str], ctx: List[Any], fctxtype: str) -> FilterFormArgs:
        ff_args = await FilterFormArgs.create(plugin_ctx=self.plugin_ctx, maincorp=corpname, persist=True)
        ff_args.maincorp = corpname
        ff_args.pnfilter = 'p' if fctxtype in ('any', 'all') else 'n'
        ff_args.filfpos = ctx[0]
        ff_args.filtpos = ctx[1]
        ff_args.filfl = 'f' if ctx[2] > 0 else 'l'
        ff_args.inclkwic = False
        ff_args.default_attr = self.args.data.curr_default_attr_values[corpname]
        ff_args.query_type = 'advanced'
        ff_args.query = self._convert_query(attrname, items, fctxtype)
        return ff_args


class QuickFilterArgsConv:

    def __init__(self, plugin_ctx: AbstractCorpusPluginCtx, args: Args) -> None:
        self.args = args
        self.plugin_ctx = plugin_ctx

    @staticmethod
    def _parse(q: str) -> Tuple[str, ...]:
        srch = re.search(r'^([pPnN])(-?\d+)([<>]\d+)?\s(-?\d+)([<>]\d+)?\s(-?\d+)\s(.*)', q)
        if srch:
            return tuple(x.strip() if x is not None else x for x in srch.groups())
        else:
            raise ArgumentMappingError(f'Failed to parse quick filter query: {q}')

    @staticmethod
    def _incl_kwic(v: str) -> bool:
        return True if v in ('n', 'p') else False

    def __call__(self, query: str) -> FilterFormArgs:
        elms = self._parse(query)
        ff_args = FilterFormArgs(
            plugin_ctx=self.plugin_ctx,
            maincorp=self.args.maincorp if self.args.maincorp else self.args.corpname,
            persist=True)
        ff_args.data.query_type = 'advanced'
        ff_args.data.query = elms[-1]
        ff_args.data.maincorp = self.args.maincorp if self.args.maincorp else self.args.corpname
        ff_args.data.pnfilter = elms[0].lower()
        ff_args.data.filfl = elms[5]
        ff_args.data.filfpos = elms[1]
        ff_args.data.filtpos = elms[3]
        ff_args.data.inclkwic = self._incl_kwic(elms[0])
        ff_args.data.qmcase = True
        return ff_args
